Tools/Scripts/webkitpy/benchmark_runner/github_downloader.py

```diff
@@ -0,0 +1,195 @@
+                remaining_retry -= 1
+                try:
+                    return function(*args, **kwargs)
+                except Exception as error:
+                    _log.warning(f'Encountered {error} in {function.__name__}, remaining retry: {remaining_retry}.')
+                    if not remaining_retry:
+                        raise
+        return wrapped
+    return decorator
+
+
+class GithubDownloader(object):
+    def __init__(self, url, owner, repository):
+        self._url = url
+        self._owner = owner
+        self._repository = repository
+        self._session = None
+
+    @classmethod
+    def is_directory(cls, entry):
+        return entry['mode'].startswith('040')
+
+    @classmethod
+    def is_file(cls, entry):
+        return entry['type'] == 'blob' and not cls.is_symlink(entry)
+
+    @classmethod
+    def is_symlink(cls, entry):
+        return entry['mode'].startswith('120')
+
+    @property
+    def session(self):
+        if self._session is None:
+            self._session = requests.Session()
+        return self._session
+
+    def download_content(self, relpath_in_repo, commit_hash, destination, subtree=None):
+        relpath_in_repo = os.path.normpath(relpath_in_repo)
+        # "subtree" can be None, a subtree url or a cached and compressed http response from subtree API.
+        # Latter two are intended to reduce Github api call.
+        # The second will issue one API request, and the last will NOT issue any API request.
+        # Unauthenticated user will be limited to 60 requests per hour.
+        # https://docs.github.com/en/rest/rate-limit#about-the-rate-limit-api
+        if subtree is None:
+            subtree = self.subtree_for_path_with_commit(relpath_in_repo, commit_hash)
+        elif not isinstance(subtree, dict):
+            subtree = self._subtree(subtree, recursive=True)
+        return self._download_subtree_multiprocessing(relpath_in_repo, commit_hash, subtree, destination)
+
+    def subtree_for_path_with_commit(self, relpath_in_repo, commit_hash):
+        commit = self._commit(commit_hash)
+        subtree = self._subtree(commit['tree']['url'])
+        parts = relpath_in_repo.split(os.path.sep)
+        last_index = len(parts) - 1
+        visited_path = []
+        for index, part in enumerate(parts):
+            for tree in subtree['tree']:
+                if tree['path'] == part:
+                    subtree = self._subtree(tree['url'], recursive=(index == last_index))
+                    visited_path.append(part)
+                    break
+            else:
+                raise Exception(f'Path {part} does not exists under {os.path.join(*visited_path)}.')
+        return subtree
+
+    def _download_subtree_multiprocessing(self, base_relpath_in_repo, commit_hash, subtree, output_directory):
+        _log.info(f'Checking out content under {base_relpath_in_repo}:')
+        result_by_path = {}
+
+        with Pool(10) as pool:
```

Comment:
hmm our Python2 implementation purposely had a broken implementation of Pool.  Thats why we don't use it anywhere.  Not sure if anything changed for Python3 so using this is risky

Reply:
Per discussion, python3 uses `spawn` on macOS by default since python3.8.
https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods